src/main/java/org/folio/rest/service/AccountService.java

```diff
@@ -0,0 +1,74 @@
+package org.folio.rest.service;
+
+import static io.vertx.core.Future.succeededFuture;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+import static org.folio.HttpStatus.HTTP_NO_CONTENT;
+import static org.folio.rest.domain.FeeFineStatus.CLOSED;
+import static org.folio.rest.domain.FeeFineStatus.forValue;
+import static org.folio.rest.jaxrs.resource.Accounts.PutAccountsByAccountIdResponse;
+import static org.folio.rest.jaxrs.resource.Accounts.PutAccountsByAccountIdResponse.respond500WithTextPlain;
+import static org.folio.rest.persist.PgUtil.put;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+import javax.ws.rs.core.Response;
+
+import org.apache.commons.lang3.StringUtils;
+import org.folio.rest.domain.FeeFineStatus;
+import org.folio.rest.domain.FeeFineAmount;
+import org.folio.rest.jaxrs.model.Account;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Context;
+
+public class AccountService {
+  private static final Logger log = LoggerFactory.getLogger(AccountService.class);
+  private static final String ACCOUNTS_TABLE = "accounts";
+
+  public CompletableFuture<AsyncResult<Response>> updateAccount(String accountId,
+    Account account, Map<String, String> headers, Context context) {
+
+    final PubSubService pubSubService = new PubSubService(headers, context);
+    final CompletableFuture<AsyncResult<Response>> putCompleted = new CompletableFuture<>();
+
+    put(ACCOUNTS_TABLE, account, accountId, headers, context,
+      PutAccountsByAccountIdResponse.class, putCompleted::complete);
+
+    return putCompleted.thenCompose(responseResult -> {
+      if (!isFeeFineUpdateSucceeded(responseResult)) {
+        return completedFuture(responseResult);
+      }
+
+      pubSubService.publishAccountBalanceChangeEvent(account);
+
+      if (isFeeFineWithLoanClosed(account)) {
+        return pubSubService.publishFeeFineWithLoanClosedEvent(account)
+          .thenApply(notUsed -> responseResult);
+      }
+
+      return completedFuture(responseResult);
+    }).exceptionally(error -> {
+      log.error("Cannot publish fee/fine closed event [loanId - {}, feeFineId - {}]," +
+        " error occurred {}", account.getLoanId(), account.getId(), error);
+
+      return succeededFuture(respond500WithTextPlain(error.getMessage()));
```

Comment:
Correct me if I'm wrong, but if the update of an account in the DB was successful, but the event publishing failed, we would still return 500, right? Is this the desired behavior?

Reply:
FOLIO does not yet have guidance on what should happen when a process can complete except for the publishing of events. There is a group working on this.

In the mean time, I think we agreed to fail so that the user would at least be aware that any follow on process will not happen.

Reply:
What is this group? Is there a way for us to track progress of their work?

Reply:
The group is made up of Taras, Craig, Vince, Jakub and myself at the moment. There is the outline of a working document (I've haven't had chance to update it for a week or so) - https://wiki.folio.org/display/DD/Handling+Failure+in+Asynchronous+Processes

It is likely this group will report progress to the Technical Leads and the Technical Council as well.

Reply:
@marcjohnson-kint thank you